riot.py

```diff
@@ -0,0 +1,480 @@
+  for these and copy it
+- does specifying encoding="utf-8" affect the locale of the test cases?
+- reporting to datadog (stats, traces)
+- typing
+- clean up into classes w/ methods?
+- better log formatter / output
+- better handling of CmdFailure in run_suites
+- add command to run pytest (or any command) in a venv
+  eg: python riot.py --suite tracer --python 3.7 --command pytest -k test_tracer_fork
+"""
+import argparse
+import itertools
+import logging
+import os
+import re
+import shutil
+import subprocess
+import sys
+import typing as t
+
+
+logger = logging.getLogger(__name__)
+
+
+SHELL = "/bin/bash"
+ENCODING = "utf-8"
+
+
+class AttrDict(dict):
+    def __init__(self, *args, **kwargs):
+        super(AttrDict, self).__init__(*args, **kwargs)
+        self.__dict__ = self
+
+
+class CmdFailure(Exception):
+    def __init__(self, msg, completed_proc):
+        self.msg = msg
+        self.proc = completed_proc
+        self.code = completed_proc.returncode
+        super().__init__(self, msg)
+
+
+# It's easier to read `Suite`, `Case` rather than AttrDict or dict.
+CaseInstance = Case = Suite = AttrDict
+
+global_deps = [
+    "mock",
+    "opentracing",
+    "pytest<4",
+    "pytest-benchmark",
+    "pytest-django",
+]
+
+global_env = [("PYTEST_ADDOPTS", "--color=yes")]
+
+all_suites = [
+    Suite(
+        name="tracer",
+        command="pytest tests/test_tracer.py",
+        env=dict(),
+        cases=[
+            Case(pys=[2.7, 3.5, 3.6, 3.7, 3.8,], pkgs=[("msgpack", [None, "==0.5.0", ">=0.5,<0.6", ">=0.6.0,<1.0"])],),
+        ],
+    ),
+    Suite(
+        name="redis",
+        command="pytest tests/contrib/redis/",
+        cases=[
+            Case(
+                pys=[2.7, 3.5, 3.6, 3.7, 3.8,],
+                pkgs=[("redis", [">=2.10,<2.11", ">=3.0,<3.1", ">=3.2,<3.3", ">=3.4,<3.5", ">=3.5,<3.6"])],
+            ),
+        ],
+    ),
+    Suite(
+        name="profiling",
+        command="python -m tests.profiling.run pytest --capture=no --verbose tests/profiling/",
+        env=[("DD_PROFILE_TEST_GEVENT", lambda c: "1" if "gevent" in c.pkgs else None),],
+        cases=[
+            Case(pys=[2.7, 3.5, 3.6, 3.7, 3.8], pkgs=[("gevent", [None, ""])],),
+            # Min reqs tests
+            Case(pys=[2.7], pkgs=[("gevent", ["==1.1.0"]), ("protobuf", ["==3.0.0"]), ("tenacity", ["==5.0.1"]),]),
+            Case(
+                pys=[3.5, 3.6, 3.7, 3.8],
+                pkgs=[("gevent", ["==1.4.0"]), ("protobuf", ["==3.0.0"]), ("tenacity", ["==5.0.1"]),],
+            ),
+        ],
+    ),
+    Suite(
+        name="django",
+        command="pytest tests/contrib/django",
+        cases=[
+            Case(
+                env=[("TEST_DATADOG_DJANGO_MIGRATION", [None, "1"])],
+                pys=[2.7, 3.5, 3.6],
+                pkgs=[
+                    ("django-pylibmc", [">=0.6,<0.7"]),
+                    ("django-redis", [">=4.5,<4.6"]),
+                    ("pylibmc", [""]),
```

Comment:
I wonder if we should alias `"latest" -> ""` would it make it more clear to read?

Reply:
or just `latest = ""` -> ` ("pylibmc", [latest])`. Either way I think we're in agreement that "" is kinda subtle.